cpp/tools/parquet/CMakeLists.txt

```diff
@@ -32,5 +32,16 @@ if(PARQUET_BUILD_EXECUTABLES)
             RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR})
   endforeach(TOOL)
 
+  # Only build parquet-dump-footer when statically linking.
```

Comment:
Hmm, why is that required? Because we're accessing Thrift internals perhaps? In that case, I would recommend exposing the necessary Thrift-accessing functions as internal APIs from libparquet, and remove the static linking requirement.

Reply:
This is because thrift doesn't `__declspec(dllexport)`/`__declspec(dllimport)` in the generated `parquet_types.h` header. Do we want to redeclare those one by one?

Reply:
No, as I said, you can instead Thrift-manipulating functions from this executable into libparquet as internal APIs. Then you don't need to access Thrift internals here.

Reply:
So pull `Scrub` into libparquet and use it here? And perhaps the serialize/deserialize functions if those are an issue as well?

Reply:
So we are committing to `Scrub`(and other APIs) becoming a public API of `libparquet`.

For my education, what are we getting by shipping dynamically linked tools to justify the above commitment? Is there a requirement for them in some platform. Perhaps it is best to always ship statically linked tools instead?

Reply:
Btw I am trying this and it seems that the main issue is that `FileMetaData` is not exposed. Then this becomes pretty inconvenient? `Scrub` can't be a public API because it takes a pointer to a thing that is not exported. Plus I would need to add a serialization API to binary and one for json.

Maybe I am missing something? If you could sketch out how this will look I can give it a shot.

Reply:
> So we are committing to `Scrub`(and other APIs) becoming a public API of `libparquet`.

Not really, my proposal is that they are exposed as internal APIs. We have other examples of this, for example the fuzzing entrypoint:
https://github.com/apache/arrow/blob/0bae073d1abe439d113cc12e06e7ada886a2f2fd/cpp/src/parquet/arrow/reader.h#L372-L377

For example, you could instantiate a `FileMetaData` object from the serialized metadata, and implement a `FileMetaData::ScrubSensitiveData` method that does what you want. `FileMetaData` has a `SerializeToString` method that reencodes to Thrift, and you could add another `SerializeToJson` method.

Or, if `FileMetaData` does not work, you can just move the current helper functions into libparquet, taking care of replacing `format::FileMetaData` with either `std::any` or a dedicated opaque wrapper. Not particularly pretty, but it would certainly work.

> For my education, what are we getting by shipping dynamically linked tools to justify the above commitment? Is there a requirement for them in some platform? Perhaps it is best to always ship statically linked tools instead?

It's not even about shipping the tools, it's about getting working tools on developers' workstation without having to be forced into static linking. I'm also not sure static linking works properly on Windows... and it tends to blow up file size a lot.

Reply:
> For example, you could instantiate a FileMetaData object from the serialized metadata, and implement a FileMetaData::ScrubSensitiveData method that does what you want. FileMetaData has a SerializeToString method that reencodes to Thrift, and you could add another SerializeToJson method.

Ah now I see. You want to use `FileMetaData` and not `format::FileMetaData`. I opted from not exposing `Scrub` because it is much harder to implement if we want to preserve `FileMetaDataImpl` invariants - `schema_` would get out of sync with the `metadata_`. I opted from a function that serializes the metadata unencrypted and takes options to enable scrubbing and/or json.

> It's not even about shipping the tools, it's about getting working tools on developers' workstation without having to be forced into static linking. I'm also not sure static linking works properly on Windows... and it tends to blow up file size a lot.

FWIW the size of this particular statically linked executable is 200kb. Unfortunately I don't have a windows machine to test it there how large it would be... 🤷

Reply:
> FWIW the size of this particular statically linked executable is 200kb.

Yeah, I'm thinking about the entire build size (including, potentially unit tests, etc.).

Reply:
> Yeah, I'm thinking about the entire build size (including, potentially unit tests, etc.).

Yeah if we were forcing static linking for unit tests that would add to build time substantially. In this case though it was for a leaf binary so perhaps it would have been ok.

In any case, moving the API inside libparquet solved this issue so now we can link both static and dynamic!